pkg/expression/collation.go

```diff
@@ -302,6 +302,19 @@ func deriveCollation(ctx BuildContext, funcName string, args []Expression, retTy
 
 // CheckAndDeriveCollationFromExprs derives collation information from these expressions, return error if derives collation error.
 func CheckAndDeriveCollationFromExprs(ctx BuildContext, funcName string, evalType types.EvalType, args ...Expression) (et *ExprCollation, err error) {
+	defer func() {
+		if err != nil && !collate.NewCollationEnabled() {
+			// args[0] can't be nil here.
+			// Use the collation of the first argument instead of returning an error.
+			et = &ExprCollation{
```

Comment:
It'll cause some behavior change :thinking: . For example:

```sql
select convert('1' using utf8) collate utf8_bin = convert('1' using utf8) collate utf8_general_ci;
```

It'll return an error before this PR no matter whether `new_collation` is enabled or not, but after this PR, it'll return without any error.

For this issue, I propose two other options, but none of them is perfect :facepalm: :

1. Avoid constant propogation in this case. We can replace the `NewFunctionInternal` with `NewFunction` in the `ColumnSubstituteImpl`, and check the `err`. If it returns an err, we can avoid `substitute` in this case.

2. Add back the `newExpr.SetCoercibility(v.Coercibility())` removed in https://github.com/pingcap/tidb/pull/38826. It sounds better because actually the coercibility (after `substitute`) should  be the `Column` (2). I'm not sure 100% sure, but maybe we can fix the issue found in #38826 by `Clone` the `newExpr` before `SetCoercibility`.

Reply:
The proposal 2 is not correct, because the following case will fail:

```
create table t (a char(10) collate utf8mb4_bin, b char(10) collate utf8mb4_general_ci);
insert into t values ('a', 'A');
select * from t t1, t t2 where t1.a=t2.b and t2.b='a' collate utf8mb4_general_ci;
```